pairing/bn256/point.go

```diff
@@ -435,6 +435,64 @@ func (p *pointG2) String() string {
 	return "bn256.G2:" + p.g.String()
 }
 
+func (p *pointG2) Hash(m []byte) kyber.Point {
+	// Hash the data to get the "real" and "imaginary" parts of the
+	// initial attempt at an x coordinate.  The real component
+	// becomes SHA256("r" || m) and the imaginary component becomes
+	// SHA256("i" || m).
+	x := &gfP2{*newGFp(0), *newGFp(0)}
+	realh := sha256.New()
+	realh.Write([]byte("r"))
+	realh.Write(m)
+	realhash := realh.Sum(nil)
+	imagh := sha256.New()
+	imagh.Write([]byte("i"))
+	imagh.Write(m)
+	imaghash := imagh.Sum(nil)
+	x.x.Unmarshal(imaghash[:])
+	x.y.Unmarshal(realhash[:])
+	montEncode(&x.x, &x.x)
+	montEncode(&x.y, &x.y)
+
+	// See if there's a corresponding y coordinate for this x.  If not,
+	// increment (the "real" part of) x and keep trying.
+	y := &gfP2{}
+	one := &gfP2{*newGFp(0), *newGFp(1)}
+
+	for {
```

Comment:
This same algorithm is already implemented in hashToPoint(), on line 226. Can you explain why this second implementation is needed for g2?

hashToPoint says, "ideally we want to do this using gfP, but gfP doesn't have a ModSqrt function". In this PR you are adding a Sort to gfp2. I don't understand why g1 and g2 need different approaches to the same problem.

Reply:
By the way: I am not advocating that things should be all perfectly cleaned up and every single line of duplicate code removed. And the behaviour of hash to a G1 point must not be changed, or else the existing BLS signatures would all stop validating.

So if the explanation is, "we knew about hashToPoint and we know we can't touch it, but for X reason G2 needs the same thing but slightly different", then that's ok with me.

Reply:
The core of the hash algorithm is indeed (intentionally) the same for G1 and G2, but there are a few differences:
  - They operate over different fields (gfP vs gfP2) and groups (G1 vs the twist group that contains G2), of course
  - The G2 version has to multiply by the G2 cofactor at the end
  - The initial x coordinate (in gfP for G1 and in gfP2 for G2) is a hash of the input data, but twice as much hash output has to be produced for G2, since gfP2 elements are twice as large

So it's possible that someone could implement Sqrt for gfP, and then make common interfaces for both the fields and the groups that unify the differences in the core of the algorithm, but still initialize the initial x coordinate differently, and multiply by the cofactor only in the G2 version?  That would certainly clean up the existing G1 Hash() function that has to convert everything to BigInt and back because of a lack of Sqrt in gfP, but that should not be part of this PR.